python/lsst/pipe/drivers/singleFrameDriver.py

```diff
@@ -0,0 +1,102 @@
+import sys
+import math
+import collections
+
+import lsst.daf.base as dafBase
+import lsst.afw.table as afwTable
+import lsst.afw.image as afwImage
+from .utils import ButlerTaskRunner, getDataRef
+from lsst.pipe.base import Struct, ArgumentParser
+from lsst.pipe.tasks.processCcd import ProcessCcdTask
+from lsst.pex.config import Config, Field, ConfigurableField, ListField
+from lsst.ctrl.pool.pool import abortOnError, NODE, Pool, Debugger
+from lsst.ctrl.pool.parallel import BatchPoolTask
+
+Debugger().enabled = True
+
+
+class SingleFrameDriverConfig(Config):
+    processCcd = ConfigurableField(target=ProcessCcdTask, doc="CCD processing task")
+    ignoreCcdList = ListField(dtype=int, default=[], doc="List of CCDs to ignore when processing")
+
+class SingleFrameDriverTask(BatchPoolTask):
+    """Process an entire exposure at once.
+
+    We use MPI to gather the match lists for exposure-wide astrometric and
+    photometric solutions.  Note that because of this, different nodes
+    see different parts of the code.
+    """
+
+    RunnerClass = ButlerTaskRunner
+    ConfigClass = SingleFrameDriverConfig
+    _DefaultName = "singleFrameDriver"
+
+    def __init__(self, *args, **kwargs):
+        """Constructor.
+
+        All nodes execute this method.
+        """
+        super(SingleFrameDriverTask, self).__init__(*args, **kwargs)
+        self.makeSubtask("processCcd")
+
+    @classmethod
+    def batchWallTime(cls, time, parsedCmd, numCpus):
+        numCcds = sum(1 for raft in parsedCmd.butler.get("camera") for ccd in raft)
+        numCycles = int(math.ceil(numCcds/float(numCpus)))
+        numExps = len(cls.RunnerClass.getTargetList(parsedCmd))
+        return time*numExps*numCycles
+
+    @classmethod
+    def _makeArgumentParser(cls, *args, **kwargs):
+        doBatch = kwargs.pop("doBatch", False)
+        parser = ArgumentParser(name="singleFrameDriver", *args, **kwargs)
+        parser.add_id_argument("--id", datasetType="raw", level="visit",
+                               help="data ID, e.g. --id visit=12345")
+        return parser
+
+    @abortOnError
+    def run(self, expRef, butler):
+        """Process a single exposure, with scatter-gather-scatter using MPI.
+
+        All nodes execute this method, though the master and slaves have different
+        routes through it.  The expRef is only a DummyDataRef on the slaves.
```

Comment:
I could not find the test for DummyDataRef. It may be worth giving a hint to the reader as to where this occurs.